app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt

```diff
@@ -1,32 +1,64 @@
 package com.lightricks.feedexercise.ui.feed
 
+import android.util.Log
 import androidx.lifecycle.*
 import com.lightricks.feedexercise.data.FeedItem
+import com.lightricks.feedexercise.data.FeedRepository
+import com.lightricks.feedexercise.data.toFeedItems
 import com.lightricks.feedexercise.util.Event
+import io.reactivex.disposables.CompositeDisposable
+import io.reactivex.disposables.Disposable
+import io.reactivex.schedulers.Schedulers
 import java.lang.IllegalArgumentException
 
 /**
  * This view model manages the data for [FeedFragment].
  */
-open class FeedViewModel : ViewModel() {
-    private val isLoading = MutableLiveData<Boolean>()
-    private val isEmpty = MutableLiveData<Boolean>()
-    private val feedItems = MediatorLiveData<List<FeedItem>>()
+open class FeedViewModel(private val feedRepository: FeedRepository) : ViewModel() {
+
+    private val disposable = CompositeDisposable()
+
+    private val isLoading = MutableLiveData<Boolean>(false)
+    private val isEmpty = MediatorLiveData<Boolean>().apply {
+        this.addSource(feedRepository.getFeedItems()) { items ->
+            this.postValue(items.isEmpty())
+        }
+    }
+
     private val networkErrorEvent = MutableLiveData<Event<String>>()
 
     fun getIsLoading(): LiveData<Boolean> = isLoading
     fun getIsEmpty(): LiveData<Boolean> = isEmpty
-    fun getFeedItems(): LiveData<List<FeedItem>> = feedItems
+    fun getFeedItems(): LiveData<List<FeedItem>> =
+        Transformations.map(feedRepository.getFeedItems()) {
+            it.toFeedItems()
+        }
+
     fun getNetworkErrorEvent(): LiveData<Event<String>> = networkErrorEvent
 
     init {
         refresh()
     }
 
     fun refresh() {
-        //todo: fix the implementation
-        isLoading.value = false
-        isEmpty.value = true
+        isLoading.value = true
+        disposable.add(feedRepository.refresh().subscribeOn(Schedulers.io())
+            .observeOn(Schedulers.io())
```

Comment:
Well, I vaguely assumed that because I'm doing db transactions here, it's better to be on a different thread, but I guess I was wrong and the observing stream better to stay on the UI thread.